### ISSUE TITLE
New attribute - "event"?

### ISSUE THREAD

**Issue by samscott89** [tokio-rs/tracing#422]
### Crates

`tracing-attributes`

### Motivation

It would be nice to be able to create events on invoking a method, much like how `#[instrument]` can create a span for a method.

The rationale is that I find myself often picking between:
```rust
// (1)
#[instrument]
fn do_thing(x: X, y: Y) {
    trace!("doing a thing");
}
```
and
```rust
// (2)
fn do_thing(x: X, y: Y) {
    trace!({ ?x, ?y}, "do_thing");
}
```
where the former is more convenient to write, but ends up with a redundant "doing a thing" message.

An alternative might be:

```rust
// (3)
#[instrument]
fn do_thing(x: X, y: Y) {
    trace!("call");
}
```

but this ends up littering code with unhelpful noise.

### Proposal

Add a new `#[event]` proc macro, which does virtually the same as `#[instrument]` but outputs an event instead. The above would just be:
```rust
#[event]
fn do_thing(x: X, y: Y) {
   // ...
}
```

which would expand to (2) above.

Presumably the ordering of attributes here would be somewhat important, if somebody for whatever reason wanted to do both `#[event]` and `#[instrument]`.

### Alternatives

This is a fairly minor nit, so leaving it alone is a perfectly reasonable option.

**Comment by samscott89** [comment]
Example implementation almost 1 for 1 copy of `instrument`: https://github.com/tokio-rs/tracing/compare/master...samscott89:sam/event-attribute

**Comment by hawkw** [comment]
Hmm, I'm open to something like this, if folks think it would be useful. I do have a couple things to add though:

1. If we want to re-export this attribute from the `tracing` crate, the way we do with `#[instrument]`, it can't be called `#[event]`. Unfortunately, attribute macros and function-like macros occupy the same namespace, so that would conflict with the `tracing::event!` macro.
2. Another alternative would be adding an option to the `#[instrument]` attribute that generates an event as well as a span, something like

   ```rust
   #[instrument(event = true)]
   fn do_thing(x: X, y: Y) {
       // ...
   }
   ```
   This would have the advantage of inherently avoiding the problem of attribute ordering you mentioned, as well as avoiding conflicts with the name of the existing `event!` macro, but, on the other hand, it might be slightly less discoverable/ergonomic.

What do folks think? @LucioFranco @davidbarsky any opinions?

**Comment by davidbarsky** [comment]
I might be misunderstanding the _reason_ for opening this issue—please correct me if I'm wrong, @samscott89.

A while ago, I believe @yaahc, @hawkw, and I were discussing introducing an option to `FmtSubscriber` where it'd write out the current span (and their parental hierarchy) each time a span would enter and exit. Currently, `FmtSubscriber` only writes that data if an event occurred in that span. However, I think that can be generalized to support writing under _any_ callback exposed by [`Layer`](https://tracing.rs/tracing_subscriber/layer/trait.layer#provided-methods). Would a toggle that sets the conditions under which `FmtSubscriber` (via `fmt::Layer`) emits a span's data be an acceptable alternative to adding a new macro `event` macro?

**Comment by hawkw** [comment]
(for those following along from home, the feature @davidbarsky's referring to is tracked in issue https://github.com/tokio-rs/tracing/issues/161)

**Comment by yaahc** [comment]
I wanted a setting in the subscriber for being able to add implicit `enter` and or `exit` events upon entering / exiting spans. If we were to do this as an attribute I think I'd prefer if it was independent of the intstrument macro. As in you have one macro for adding spans to a function, and another macro for adding entry and or exit events to a function. That and a global setting on the fmt subscriber to implicitly turn on even / exit logging temporarily on all spans would scratch every itch I have (I think).

**Comment by hawkw** [comment]
@yaahc I don't really understand the use-case for having _both_ an attribute like the one you're describing, _and_ adding a configuration to `FmtSubscriber` to log enter/exit. Is there a reason we would need both? This seems like a whole lot of options for doing more or less the same thing, and I'm not sure it's worth presenting users with a lot of choices that are similar but subtly different, unless there's a specific need that's only met by multiple options...

Also, what would the `enter`, `exit`, `open` and `close` arguments in your sample do? Would this emit an event when the _span_ generated by `#[instrument]` is entered, exited, opened, and closed? To me, this seems superfluous — the subscriber is already notified when those things happen. Or, is the `#[event]` attribute modifying the _function_? In that case, what does "open" and "close" mean for a function?

**Comment by yaahc** [comment]
@hawkw The usecase for having both is that presumably the global one could be controlled by an env variable or something so that you can turn it on without recompiling. It could then be used like trace logging where you have your subscriber just start logging events for every span enter/exit/open/close so you can just turn on the firehose and quickly identify what your problem is without having to instrument a ton of your program. The `#[event]` attribute would be for more permanent situations on a per span/function basis.

As for the enter exit open closer, if you could somehow configure these events at a per span level that would be like inserting the above event logging from the subscriber but only enabling it for a specific span. If the attribute is just for modifying the function then yea, enter and exit are the only two that make sense, as I mentioned at the end of my previous comment.

**Comment by hawkw** [comment]
@yaahc I guess I just don't see a compelling _reason_ to emit an `Event` on span enter, exit, open, and close. We already _have_ subscriber methods for these events (in the little-'e' sense, not the type) that are more semantically meaningful within `tracing` than an `Event` with a message like "span entered". 

I don't think adding `Event`s to these span lifecycle events at the instrumentation point is a good idea. These are _not_ user invocations of the `event!` macro; they mean something different. If we put `Event`s in enter, exit, open, and close at the instrumentation point, then those events are now being emitted, no matter how the `Subscriber` chooses to handle them. If users want to handle these span events identically to how they handle an `Event`, that's fine — the Right Thing to do then would be to implement a `Subscriber` that handles the `Subscriber::{enter, exit, open, close}` methods the same way it handles the `Event` method.

**Comment by yaahc** [comment]
>  the Right Thing to do then would be to implement a Subscriber that handles the Subscriber::{enter, exit, open, close} methods the same way it handles the Event method.

This is the original proposal from #161 right? Where you have settings on the subscriber to enable these various Event-like handling blocks within the `Subscriber::{enter, exit, open, close}` methods.

**Comment by hawkw** [comment]
Right, we want the `fmt` subscriber to have the option to log span lifecycle events. Other subscribers could do similar things if it makes sense with the behavior they implement

**Comment by samscott89** [comment]
Thanks all for adding a bunch of stuff to this thread I really hadn't even thought of.

The original motivation was mostly inspired by laziness. Basically, there's a lovely `instrument` macro which captures a bunch of variables, and (assuming use of a subscriber like the `fmt` one), I can easily see all of these variables if I create an event right after.

This probably isn't the right approach to this, however, since it's conflating the meanings of spans/events and is primarily just about getting some variables logged.

And so I don't think about this in the same context as using things like enter/exit span.

I _think_ the overhead of adding a new macro for just this feature doesn't quite seem worth it. In most cases where I was using it, when I tried this out locally using the implementation I linked, it turned out there was a more meaningful event somewhere else that I could be creating, instead of adding a vacuous trace to the top of the method.

So I'm happy to close this for that part of it. If the discussion seems useful around other variants of this macro, feel free to keep it open.

**Comment by hawkw** [comment]
Closing this in favour of getting span lifecycles into `tracing-subscriber`'s `fmt`. Thanks for the discussion & input, everyone!
